### PR TITLE
[Merged by Bors] - feat(GroupTheory): covering a group by cosets (Lemma of B. H. Neumann)

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Lemma of B. H. Neumann on covering a group by finitely many cosets of subgroups.

---

[Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/generic.20great.20circle.20on.20the.20sphere.20misses.20any.20finite.20set/near/439361300)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by alreadydone — APPROVED** [review]
Thanks for working on this! In addition to the suggestions I have some golfs [here](https://github.com/leanprover-community/mathlib4/commit/bfdc21d3a536beba5866bda6bdfae4287ffdef2d). LGTM otherwise

File: Mathlib/GroupTheory/Complement.lean [hunk 1]
```diff
@@ -580,6 +588,15 @@ noncomputable def toEquiv (hS : S ∈ Subgroup.rightTransversals (H : Set G)) :
...
 #align add_subgroup.mem_right_transversals.to_equiv AddSubgroup.MemRightTransversals.toEquiv
 
+@[to_additive "A right transversal is finite iff the subgroup has finite index"]
+theorem finite_iff
+    (h : S ∈ Subgroup.rightTransversals (H : Set G)) :
+    Finite S ↔ H.FiniteIndex := by
+  rw [← Equiv.finite_iff (Subgroup.MemRightTransversals.toEquiv h)]
+  rw [Equiv.finite_iff (QuotientGroup.quotientRightRelEquivQuotientLeftRel H)]
+  exact ⟨fun _ ↦ finiteIndex_of_finite_quotient H,
+    fun _ ↦ finite_quotient_of_finiteIndex H⟩
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Complement.lean:598]:
```suggestion
  rw [← (Subgroup.MemRightTransversals.toEquiv h).finite_iff,
    (QuotientGroup.quotientRightRelEquivQuotientLeftRel H).finite_iff]
  exact ⟨fun _ ↦ finiteIndex_of_finite_quotient H, fun _ ↦ finite_quotient_of_finiteIndex H⟩
```

File: Mathlib/GroupTheory/Complement.lean [hunk 2]
```diff
@@ -530,6 +530,14 @@ noncomputable def toEquiv (hS : S ∈ Subgroup.leftTransversals (H : Set G)) : G
 #align subgroup.mem_left_transversals.to_equiv Subgroup.MemLeftTransversals.toEquiv
 #align add_subgroup.mem_left_transversals.to_equiv AddSubgroup.MemLeftTransversals.toEquiv
 
+@[to_additive "A left transversal is finite iff the subgroup has finite index"]
+theorem finite_iff
+    (h : S ∈ Subgroup.leftTransversals (H : Set G)) :
+    Finite S ↔ H.FiniteIndex := by
+  rw [← Equiv.finite_iff (Subgroup.MemLeftTransversals.toEquiv h)]
+  exact ⟨fun _ ↦ finiteIndex_of_finite_quotient H,
+    fun _ ↦ finite_quotient_of_finiteIndex H⟩
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Complement.lean:539]:
```suggestion
  rw [← (Subgroup.MemLeftTransversals.toEquiv h).finite_iff]
  exact ⟨fun _ ↦ finiteIndex_of_finite_quotient H, fun _ ↦ finite_quotient_of_finiteIndex H⟩
```

**Comment by bustercopley** [comment]
Thanks for the review, @alreadydone. I'll have a go at the corollary you stated, but probably nothing stronger for now.

**Comment by bustercopley** [comment]
Do you have any opinion on where `Subspace.union_ne_univ_of_lt_top` should go? For my part I think it's ok here in `GroupTheory.CosetCover`, as a corollary. An alternative is `LinearAlgebra.Dimension.DivisionRing`. (`#find_home` says `Algebra.CharP.ExpChar`, which doesn't sound good.)
```
variable {k E ι : Type*} [DivisionRing k] [Infinite k] [AddCommGroup E] [Module k E]

theorem Subspace.union_ne_univ_of_lt_top (s : Finset (Subspace k E)) (hs : ∀ p ∈ s, p < ⊤) :
    ⋃ p ∈ s, (p : Set E) ≠ Set.univ := by
```

**Comment by AntoineChambert-Loir** [comment]
I believe it sits nicely here. (Moreover, putting it elsewhere would require to add an import to that file.)

**Comment by alreadydone** [comment]
Thanks! [Here](https://github.com/leanprover-community/mathlib4/commit/ab3feb84d91f43a7ec55ef93a28e396e71e907b3) is some slight golf and a rephrase of the vector space result:
```
theorem Subspace.exists_eq_top_of_biUnion_eq_univ (hcovers : ⋃ p ∈ s, (p : Set E) = Set.univ) :
    ∃ p ∈ s, p = ⊤
```
I also changed `p < ⊤` in the assumptions to `p ≠ ⊤` for the same reason that people prefers `n ≠ 0` in the assumptions but `0 < n` in the conclusions for `n : ℕ`.

I'm not completely sure what should be the canonical spelling regarding finiteness. We've chosen to use Finset but users may apply it to a finite type (`Finite`) or a finite set (`Set.Finite`). However it's not too hard to convert between them and the current choice is reasonable, and I don't think we should provide three versions of every result.

One result in this PR can be strengthened to say that if the sum (over the finite-index cosets) of 1/index equals 1, then the finite-index cosets must be disjoint. This could be used to show that #K+1 proper vector subspaces are necessary to cover a vector space over a finite field K. (In finite dimension one can also use counting. The argument in [Clark's paper](https://arxiv.org/pdf/1208.0975) uses the irredundant result which may not be of enough general interest; but I think it's worth adding the results saying #K+1 vector spaces or #K affine subspaces are necessary to cover a finite-dimensional vector space, in some later PR.

**Comment by bustercopley** [comment]
> [Here](https://github.com/leanprover-community/mathlib4/commit/ab3feb84d91f43a7ec55ef93a28e396e71e907b3) are some slight golf and a rephrase of the vector space result

Thank you, I pushed those.

**Comment by AntoineChambert-Loir** [comment]
The pr is very close to proving disjointness of cosets when the density equals one. It might however be necessary to refactor the intermediate construction to transfer more properties than is presently done. (Density is unchanged, nor is disjointness)

For vector spaces over finite fields, I am wondering whether proofs in the Chevalley-Warning/Combinatorial Nullstellensatz style aren't easier. (Does Mathlib have the latter?)

**Comment by bustercopley** [comment]
> The pr is very close to proving disjointness of cosets when the density equals one. It might however be necessary to refactor the intermediate construction to transfer more properties than is presently done. (Density is unchanged, nor is disjointness)

[Here](https://github.com/leanprover-community/mathlib4/commit/9b1aa8eb6839b1befe7475596bb4d0276095f4a9) is a commit that includes the density and disjointness in the transfer. I haven't tried to use it to prove disjointness of cosets when the density equals one. (The commit also introduces a `structure LeftCosetCover`. It might be a bit too heavyweight.)

> For vector spaces over finite fields, I am wondering whether proofs in the Chevalley-Warning/Combinatorial Nullstellensatz style aren't easier. (Does Mathlib have the latter?)

Can you trick me into believing it's not as difficult as it sounds? :smile:

**Comment by AntoineChambert-Loir** [comment]
OMG! I was in the middle of proving disjointness. Probably if I pull your commit, it'll be done.

**Comment by AntoineChambert-Loir** [comment]
I can assure you, it is not as difficult as it sounds.
In fact, for subspaces of vector spaces, it would have been easier to use this machine than Neumann's lemma (but I didn't know that lemma and found it fun!).

**Comment by bustercopley** [comment]
> I can assure you, it is not as difficult as it sounds. In fact, for subspaces of vector spaces, it would have been easier to use this machine than Neumann's lemma (but I didn't know that lemma and found it fun!).

I was quite taken with it too! (Thanks again for the suggestion, @alreadydone.)

**Comment by bustercopley** [comment]
I had no success refactoring the unwieldy auxiliary lemma `Subgroup.leftCoset_cover_filter_FiniteIndex_aux`, I'm afraid.

**Comment by AntoineChambert-Loir** [comment]
I'll try something this afternoon once I've finished that urgent stack of grading.

File: Mathlib/GroupTheory/CosetCover.lean [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+-/
+
+open scoped Pointwise BigOperators
+
+section leftCoset_cover_const
+
+variable {G : Type*} [Group G]
+
+@[to_additive]
+theorem Subgroup.exists_leftTransversal_of_FiniteIndex
```
Comment (jcommelin) [inline Mathlib/GroupTheory/CosetCover.lean:52]: Please use a `namespace Subgroup` to shorten the names of the lemmas up to line 338.

**Review by jcommelin — COMMENTED** [review]
Sorry, need to run after only a trivial comment.

File: Mathlib/GroupTheory/CosetCover.lean (hunk 3 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/CosetCover.lean]: Done!

**Comment by faenuccio** [comment]
I have removed the `awaiting-review` label according to the decision explained here
https://leanprover.zulipchat.com/#narrow/stream/113486-announce/topic/Changes.20to.20the.20.23queue/near/450176907

File: Mathlib/GroupTheory/CosetCover.lean [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+end Submodule
+
+section Subspace
+
+variable {k E ι : Type*} [DivisionRing k] [Infinite k] [AddCommGroup E] [Module k E]
+    {s : Finset (Subspace k E)}
+
+/- A vector space over an infinite field cannot be a finite union of proper subspaces. -/
+theorem Subspace.biUnion_ne_univ_of_ne_top (hs : ∀ p ∈ s, p ≠ ⊤) :
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/CosetCover.lean:382]: Can this be generalized to free module over some class of rings?

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/CosetCover.lean:382]: It's false for Z² as a Z-module, by taking the pre images of the three mod-2 lines. Maybe if all residue fields are infinite...

Comment (riccardobrasca) [reply Mathlib/GroupTheory/CosetCover.lean:382]: OK, it's complicated. It doesn't matter.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/CosetCover.lean:382]: It's a good question, obviously, and in some sense, Neumann's lemma gives some answers. If you have a finitely generated module, then every submodule is contained in a maximal one whose co-quotient is a residue field, hence is infinite if all residue fields are infinite.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/CosetCover.lean:382]: But I have to admit that my way of proving the case of vector spaces is not via Neumann's lemma but via polynomials (over an infinite field, a nonzero polynomial does not vanish identically).

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Comment by riccardobrasca** [comment]
As usual something changed in mathlib, please merge master and fix the error. Thanks!

bors d+

**Pull request closed without merging**